Fix/#42 redis queue

bjcareer opened the pull request:
## 구조 변경
기존에는 락을 획득한 후에 캐시 데이터를 업데이트하는 방식이었습니다. 그러나, 이 방식 대신에 요청이 들어온 시간과 `userId`를 우선순위 큐에 넣는 방식으로 구조를 변경하였습니다. 
![image](https://github.com/user-attachments/assets/f3dbb69a-c64f-4315-9bbb-c20873e366e9)

그 이후에는 TPS가 아래와 같이 나오는 것을 확인했습니다.
![image](https://github.com/user-attachments/assets/5865165a-5043-4fac-9898-cfb3efc73ec4)

이후에는 scale-out을 했을 때도 TPS가 상승하는 것을 확인하기 위해서 서버를 두 대로 올려서 진행했습니다.
![image](https://github.com/user-attachments/assets/75bd921a-d58d-4802-8e3f-93578287e448)

bjcareer committed:
[infra] Use master and salve for redis

bjcareer committed:
[refactor] Extract code for entity state logging into a separate function

bjcareer committed:
[fix] Move @Transactional annotation from Service to Repository

 - 서비스 계층에서 레토지토리로 이동함 그 이유는 이슈 #43에서 설명

 related to #43

bjcareer committed:
[feat] Implement QueueService to eliminate lock contention

 - 요청마다 레디스의 락을 획득하니, TPS가 느려짐

 related to #42

bjcareer committed:
[feat] Implement QueueService to eliminate lock contention

 - 요청마다 레디스의 락을 획득하니, TPS가 느려짐

 related to #42

bjcareer committed:
[feat] User redis queue for scale-out
  - 서버 확장을 위해서 메모리 큐가 아닌 레디스 큐를 사용했습니다.

 related to #42

bjcareer committed:
[refactor] Create Redis domain

 - 레디스를 이용하는 코드를 한 곳으로 응집함

bjcareer committed:
[refactor] Save to DB when retrieving from Redis queue

- 다시 레디스에 넣어서 expire를 기다리는 방법이 아니라 디비에 저장하는 방식으로 변경함

- 데이터의 정합성을 유지하기 위해서 위 방식을 채택함

bjcareer committed:
[refactor] Refactor code
- 정합성을 더 확실하게 챙기는 방식으로 코드 변경

bjcareer committed:
[refactor] Refactor code
 - 요청 분산을 위해서 nginx 사용
 - 응답 메세지 변경

bjcareer committed:
[fix] Fix error Handler

 - redis lock을 획득하는 과정에서 InterruptedException이 발생하는 것을 확인했습니다.

 - 이를 도메인에서 처리하는 방식이 아니라 상위 계층에서 처리하도록 변경

f-lab-troy commented on timeDeal/src/main/java/com/bjcareer/stockservice/RedisConfig.java:
```diff
@@ -56,9 +62,16 @@ public Object setupKeyExpirationListener(RedissonClient redissonClient, RedisLis
+
+	@Bean
+	public RScoredSortedSet<Pair<Long, String>> setUpSortedSet(RedissonClient redissonClient) {
+		return redissonClient.getScoredSortedSet(queueName);
```
sorting 없이 구현 했을때와 성능차이점 확인

bjcareer replied:
Redis에서 제공하는 ZADD 연산의 경우 Log(N)의 시간 복잡도를 가지고 있습니다. 따라서 N이 매우 커지게 되면 성능의 이슈가 있을 수 있지만 제가 가정하고 있는 사용자 수에서는 영향이 미비할 것이라고 판단됩니다. 하지만 말씀하신 것처럼 서버의 시간 동기화등의 시스템 시간을 사용하지 않고, set의 크기를 사용해서 사용자들의 쿠폰 발급을 진행하겠습니다.

https://redis.io/docs/latest/commands/zrem/ 
https://redis.io/docs/latest/commands/zadd/
https://redis.io/docs/latest/commands/zrange/

f-lab-troy commented on timeDeal/src/main/java/com/bjcareer/stockservice/timeDeal/domain/redis/Redis.java:
```diff
@@ -0,0 +1,52 @@
+	private static final long LEASE_TIME = 1L;
+
+	public boolean acquireLock(String key) throws InterruptedException {
+		log.debug("Requesting lock acquisition for key: {}", key);
```
logging수에 따른 성능저하 확인

bjcareer replied:
로그레벨에 따라서 성능 저하가 발생하는 것을 확인했습니다. 그 이유는 로그를 콘솔에 출력하는 작업이 디스크 i/o작업으로 속도가 느리고 또한 로그의 문자열이 생성되면서 GC에 영향을 주기 때문입니다

f-lab-troy commented on timeDeal/src/main/java/com/bjcareer/stockservice/timeDeal/domain/redis/Redis.java:
```diff
@@ -0,0 +1,52 @@
+		log.debug("Retrieving single key for pattern: {}", keyPattern);
+		RKeys keys = client.getKeys();
+		Iterable<String> matchingKeys = keys.getKeysByPattern(keyPattern, 1);
+		return matchingKeys.iterator().hasNext() ? matchingKeys.iterator().next() : null;
```
iterator 문서 확인 .next()

bjcareer replied:
문서를 확인해보니 값이 없을 때 next()를 호출하게 되면 SuchElementException을 Throw하는 것을 확인했습니다!

f-lab-troy commented on timeDeal/src/main/java/com/bjcareer/stockservice/timeDeal/service/CronQueueService.java:
```diff
@@ -0,0 +1,76 @@
+
+		// RScoredSortedSet에서 데이터를 가져와서 Map에 저장
+		for (int i = 0; i < 30; i++) {
+			Double score = pq.firstScore();
```
동시성 문제 확인

bjcareer replied:
동시성 문제가 발생되는 것을 확인하여 #46 에서 그 원인과 해결법을 설명했습니다 감사합니다~

f-lab-troy commented on timeDeal/src/main/java/com/bjcareer/stockservice/timeDeal/service/CronQueueService.java:
```diff
@@ -0,0 +1,76 @@
+		}
+
+		RScoredSortedSet<String> pq = redis.getScoredSortedSet(key);
+		Map<String, Double> participationScores = new HashMap<>(30);
```
30과 같은 magic number는 상수 처리

f-lab-troy commented on timeDeal/src/main/java/com/bjcareer/stockservice/timeDeal/service/CronQueueService.java:
```diff
@@ -0,0 +1,76 @@
+		try {
+			excessCoupons = timeDealService.updateEventStatus(eventId, participationScores.size());
+		} catch (IllegalStateException e) {
+			log.error(e.getMessage());
```
로직확인

bjcareer replied:
Thread가 문제가 생겨서 Lock을 획득하지 못할 경우에는 InterruptedException을 발생하고, 대기 시간에 너무 길어져 Lock을 획득하지 못하는 경우에는 IllegalStateException을 발생시켜서 에러 처리하는 것으로 변경했습니다.

f-lab-troy commented on timeDeal/src/main/resources/application.yml:
```diff
@@ -3,7 +3,7 @@ spring:
-    url: jdbc:postgresql://localhost:5432/stock_service
+    url: jdbc:postgresql://223.130.142.94:5432/stock_service
     username: timo
     password: timo
```
password는 환경변수 또는 vault와 같은 외부서비스 사용

bjcareer replied:
문제가 되는 환경변수 노출부분은 일단 제거하고, vault연동은 TODO로 남기겠습니다~

f-lab-troy commented on timeDeal/src/test/java/com/bjcareer/stockservice/timeDeal/service/CronQueueServiceTest.java:
```diff
@@ -0,0 +1,68 @@
+	}
+
+	@Test
+	void testGetParticipationInQueue_WithLessThan30Items() {
```
중복코드 추출, 30개 테스트

f-lab-troy commented:
성능테스트시 GC 통계와 비교

bjcareer commented:
> 성능테스트시 GC 통계와 비교

related to #45를 참고 부탁드리겠습니다~

bjcareer committed:
[fix] Add concurrency handling for Redis operations

 - 레디스의 큐에서 score를 가져오는 부분에서 동시성 문제가 발생됨
 - 이를 해결하기 위해 batch를 도입함

 related to #46

bjcareer committed:
[fix] Add Return condition

 - 큐 탐색을 끝내는 조건문 추가

bjcareer committed:
[fix] Add concurrency handling for Redis operations

 - 레디스의 큐에서 size를 가지고 오는 부분에서 동시성 이슈가 발생할 가능성이 있음
 - 이를 해결하기 위해 batch를 도입함

 related to #46

bjcareer committed:
[fix] Add concurrency handling for Redis operations

- 쿠폰 발급을 검증하기 전 이벤트 객체에서 쿠폰을 발급을 업데이트 하니 중복 검증을 통해 발급 불가능해 진 쿠폰도 발급했다고 업데이트가 되는 현상이 있었음

 related to #46

bjcareer committed:
[refactor] Update Test case
